include/aws/mqtt/private/mqtt_subscription_set.h

```diff
@@ -0,0 +1,189 @@
+#include "aws/mqtt/v5/mqtt5_types.h"
+#include <aws/common/hash_table.h>
+
+struct aws_mqtt_subscription_set_subscription_options {
+    struct aws_byte_cursor topic_filter;
+
+    enum aws_mqtt5_qos qos;
+
+    bool no_local;
+    bool retain_as_published;
+    enum aws_mqtt5_retain_handling_type retain_handling_type;
+
+    aws_mqtt_client_publish_received_fn *on_publish_received;
+    aws_mqtt_userdata_cleanup_fn *on_cleanup;
+
+    void *callback_user_data;
+};
+
+struct aws_mqtt_subscription_set_subscription_record {
+    struct aws_allocator *allocator;
+    struct aws_byte_buf topic_filter;
+
+    struct aws_mqtt_subscription_set_subscription_options subscription_view;
+};
+
+struct aws_mqtt_subscription_set_publish_received_options {
+    struct aws_mqtt_client_connection *connection;
+
+    struct aws_byte_cursor topic;
+    enum aws_mqtt_qos qos;
+    bool retain;
+    bool dup;
+
+    struct aws_byte_cursor payload;
+};
+
+struct aws_mqtt_subscription_set_topic_tree_node {
+    struct aws_allocator *allocator;
+
+    struct aws_byte_cursor topic_segment_cursor; // segment can be empty
+    struct aws_byte_buf topic_segment;
+
+    struct aws_mqtt_subscription_set_topic_tree_node *parent;
+    struct aws_hash_table children; // (a node's byte_cursor -> node)
+
+    size_t ref_count;
+
+    bool is_subscription;
+
+    aws_mqtt_client_publish_received_fn *on_publish_received;
+    aws_mqtt_userdata_cleanup_fn *on_cleanup;
+
+    void *callback_user_data;
+};
+
+/**
+ * Utility type to track a client's subscriptions.
+ *
+ * The topic tree supports per-subscription callbacks as used by the MQTT311 implementation.
+ *
+ * The subscriptions table supports resubscribe APIs for both MQTT311 and MQTT5 by tracking the subscription
+ * details on a per-topic-filter basis.
+ */
+struct aws_mqtt_subscription_set {
+    struct aws_allocator *allocator;
+
+    /* a permanent ref */
+    struct aws_mqtt_subscription_set_topic_tree_node *root;
+
+    /* topic_filter_cursor -> persistent subscription */
+    struct aws_hash_table subscriptions;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/**
+ * Creates a new subscription set
+ *
+ * @param allocator allocator to use
+ * @return a new subscription set or NULL
+ */
+AWS_MQTT_API struct aws_mqtt_subscription_set *aws_mqtt_subscription_set_new(struct aws_allocator *allocator);
+
+/**
+ * Destroys a subscription set
+ *
+ * @param subscription_set subscription set to destroy
+ */
+AWS_MQTT_API void aws_mqtt_subscription_set_destroy(struct aws_mqtt_subscription_set *subscription_set);
+
+/**
+ * Checks if a topic filter exists in the subscription set's hash table of subscriptions
+ *
+ * @param subscription_set subscription set to check
+ * @param topic_filter topic filter to check for existence in the set
+ * @return true if the topic filter exists in the table of subscriptions, false otherwise
+ */
+AWS_MQTT_API bool aws_mqtt_subscription_set_is_subscribed(
+    struct aws_mqtt_subscription_set *subscription_set,
```

Comment:
Trivial - Should we be applying const to the functions that are checking but not modifying subscription sets? Could be we aren't for either testing or other reasons that I'll run into as I get through this review. Also, not entirely necessary either.

Reply:
Definitely